cameo/core/solution.py

```diff
@@ -250,26 +257,32 @@ def f(self, value):
     @property
     def fluxes(self):
         self._check_freshness()
-        primals = OrderedDict()
+        primal_values = self.model.solver.primal_values
+
+        fluxes = OrderedDict()
         for reaction in self.model.reactions:
-            primals[reaction.id] = reaction.flux
-        return primals
+            fluxes[reaction.id] = primal_values[reaction._get_forward_id()] - primal_values[reaction._get_reverse_id()]
+
+        return fluxes
 
     @property
     def reduced_costs(self):
         self._check_freshness()
-        duals = OrderedDict()
+        reduced_values = self.model.solver.reduced_costs
+
+        reduced_costs = OrderedDict()
         for reaction in self.model.reactions:
-            duals[reaction.id] = reaction.reduced_cost
-        return duals
+            reduced_costs[reaction.id] = reduced_values[reaction._get_forward_id()] - reduced_values[reaction._get_reverse_id()]
+        return reduced_costs
 
     @property
     def shadow_prices(self):
         self._check_freshness()
-        duals = OrderedDict()
+        dual_values = self.model.solver.dual_values
+        shadow_prices = {}
         for metabolite in self.model.metabolites:
-            duals[metabolite.id] = self.model.solver.constraints[metabolite.id].dual
-        return duals
+            shadow_prices[metabolite.id] = dual_values[metabolite.id]
```

Comment:
The loop is unnecessary, it could just directly return the shadow prices from the solver.